# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [5]:
import pandas as pd

# Read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4)
#print(df_rural.columns[-1])
#print(df_rural.columns[-1])

# Remove the 'Unnamed:62' column from each data set
df_rural.drop('Unnamed: 62', axis=1)
df_electricity.drop('Unnamed: 62', axis=1)

# Combine the two data sets together using the concat method
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.
df_combined = pd.concat([df_rural, df_electricity])
df_combined.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250   
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496   
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770   
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555   

      ...         2009    2010    2011    2012    2013    2014    2015  \
0     ...       56.579  56.941  57.302  57.636  57.942  58.221  58.472   
1     ...       75.687  75.311  74.926  74.532  74.129  73.718  73.297   
2     ...       60.701  59.903  59.100  58.301  57.510  56.726  55.950   
3     ...       48.924  47.837  46.753  45.670  44.617  43.591  42.593   
4     ...       11.648  12.183  12.740  13.292  13.835  14.367  14.885   

     2016    2017  Unnamed: 62  
0  58.696  58.893          NaN  
1  72.868  72.430          NaN  
2  55.181  54.422          NaN  
3  41.624  40.684          NaN  
4  15.388  15.873          NaN  

[5 rows x 63 columns]

In [6]:
df_combined.tail()

Country Name Country Code                           Indicator Name  \
259        Kosovo          XKX  Access to electricity (% of population)   
260   Yemen, Rep.          YEM  Access to electricity (% of population)   
261  South Africa          ZAF  Access to electricity (% of population)   
262        Zambia          ZMB  Access to electricity (% of population)   
263      Zimbabwe          ZWE  Access to electricity (% of population)   

     Indicator Code  1960  1961  1962  1963  1964  1965     ...       \
259  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN     ...        
260  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN     ...        
261  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN     ...        
262  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN     ...        
263  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN     ...        

          2009       2010       2011       2012       2013       2014  \
259  98.377060   0.000000  99.430435  99.773651  99.942505  99.993134   
260  61.751400  63.134281  64.532181  65.942101  75.600000  66.100000   
261  82.700000  82.900000  84.700000  85.300000  85.400000  86.000000   
262  23.235649  22.000000  24.328644  24.894667  25.469702  27.900000   
263  43.369082  36.063057  36.900000  36.728878  37.076813  32.300000   

           2015        2016  2017  Unnamed: 62  
259  100.000000  100.000000   NaN          NaN  
260   70.213905   71.642349   NaN          NaN  
261   85.500000   84.200000   NaN          NaN  
262   31.100000   27.219337   NaN          NaN  
263   33.700000   38.145138   NaN          NaN  

[5 rows x 63 columns]

# Exercise 2 (Challenge)

This exercise is more challenging.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [7]:
# Merge the data sets together according to the instructions.

# First, use the melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value

df_rural_melt = pd.melt(df_rural,
                        id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],
                        var_name = 'Year', value_name='Electricity_Value')
df_rural_melt.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code  Year  Electricity_Value  
0  SP.RUR.TOTL.ZS  1960             49.224  
1  SP.RUR.TOTL.ZS  1960             91.779  
2  SP.RUR.TOTL.ZS  1960             89.565  
3  SP.RUR.TOTL.ZS  1960             69.295  
4  SP.RUR.TOTL.ZS  1960             41.550

In [8]:
df_electricity_melt = pd.melt(df_electricity,
                              id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'],
                              var_name='Year', value_name='Rural_Value')
df_electricity_melt.head()

Country Name Country Code                           Indicator Name  \
0        Aruba          ABW  Access to electricity (% of population)   
1  Afghanistan          AFG  Access to electricity (% of population)   
2       Angola          AGO  Access to electricity (% of population)   
3      Albania          ALB  Access to electricity (% of population)   
4      Andorra          AND  Access to electricity (% of population)   

   Indicator Code  Year  Rural_Value  
0  EG.ELC.ACCS.ZS  1960          NaN  
1  EG.ELC.ACCS.ZS  1960          NaN  
2  EG.ELC.ACCS.ZS  1960          NaN  
3  EG.ELC.ACCS.ZS  1960          NaN  
4  EG.ELC.ACCS.ZS  1960          NaN

In [9]:
# Drop any columns from the data frames that aren't needed
df_rural_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)
df_electricity_melt.drop(['Indicator Name', 'Indicator Code'], axis=1, inplace=True)

# Merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year
df_combined = df_rural_melt.merge(df_electricity_melt, how='outer', on=['Country Name', 'Country Code', 'Year'])

# Sort the results by country and then by year

df_combined.sort_values(['Country Name', 'Year'], inplace=True)
df_combined.head()

Country Name Country Code  Year  Electricity_Value  Rural_Value
1     Afghanistan          AFG  1960             91.779          NaN
265   Afghanistan          AFG  1961             91.492          NaN
529   Afghanistan          AFG  1962             91.195          NaN
793   Afghanistan          AFG  1963             90.890          NaN
1057  Afghanistan          AFG  1964             90.574          NaN